src/util.rs

```diff
@@ -5,7 +5,245 @@
 #[path = "third_party/rust/layout.rs"]
 pub(crate) mod core_layout;
 
-use core::{mem, num::NonZeroUsize};
+use core::{
+    fmt::{Debug, Formatter},
+    marker::PhantomData,
+    mem,
+    num::NonZeroUsize,
+    ptr::NonNull,
+};
+
+use crate::{KnownLayout, _CastType};
+
+// For each polyfill, as soon as the corresponding feature is stable, the
+// polyfill import will be unused because method/function resolution will prefer
+// the inherent method/function over a trait method/function. Thus, we suppress
+// the `unused_imports` warning.
+//
+// See the documentation on `util::polyfills` for more information.
+#[allow(unused_imports)]
+use crate::util::polyfills::NonNullSliceExt as _;
+
+/// A raw pointer with more restrictions.
+///
+/// `Ptr<T>` is similar to `NonNull<T>`, but it is more restrictive in the
+/// following ways:
+/// - It must derive from a valid allocation
+/// - It must reference a byte range which is contained inside the allocation
+///   from which it derives
+///   - As a consequence, the byte range it references must have a size which
+///     does not overflow `isize`
+/// - It must satisfy `T`'s alignment requirement
+///
+/// Thanks to these restrictions, it is easier to prove the soundness of some
+/// operations using `Ptr`s.
+pub(crate) struct Ptr<'a, T: 'a + ?Sized> {
```

Comment:
Like `NonNull`, we should probably document the variance of the parameters. Here, both `'a` and `T` are covariant.

Reply:
Done.